File: bumperbot_vision/src/optical_flow.cpp
#include "bumperbot_vision/optical_flow.hpp"

#include "cv_bridge/cv_bridge.hpp"

using namespace std::placeholders;

namespace bumperbot_vision
{
OpticalFlow::OpticalFlow(const std::string & name)
: Node(name)
{
  approximate_sync_ = std::make_shared<ApproximateSync>(
    ApproximatePolicy(10), image_sub_, info_sub_);
  approximate_sync_->registerCallback(
    std::bind(&OpticalFlow::imageCallback, this, _1, _2));
  image_transport::TransportHints hints{this};
  const auto sensor_data_qos = rclcpp::SensorDataQoS().get_rmw_qos_profile();

  image_sub_.subscribe(
    this, "left_camera/image_raw", hints.getTransport(), sensor_data_qos);
  info_sub_.subscribe(this, "left_camera/camera_info", sensor_data_qos);

  matches_pub_ =
    image_transport::create_publisher(this, "matches", rmw_qos_profile_sensor_data);

  detector_ = cv::GFTTDetector::create(500, 0.01, 20);
}

void OpticalFlow::imageCallback(
  const sensor_msgs::msg::Image::ConstSharedPtr & image_msg,
  const sensor_msgs::msg::CameraInfo::ConstSharedPtr & info_msg)
{
  // Camera Intrinsics
  K_ = (cv::Mat_<double>(3, 3) << 
    info_msg->k[0], info_msg->k[1], info_msg->k[2],
    info_msg->k[3], info_msg->k[4], info_msg->k[5],
    info_msg->k[6], info_msg->k[7], info_msg->k[8]);

  cv::Mat img;
  try {
    img = cv_bridge::toCvCopy(image_msg, image_msg->encoding)->image;
  } catch (cv_bridge::Exception& e) {
    RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
    return;
  }

  // Convert to grayscale for better Optical Flow performance
  cv::Mat img_gray;
  if (img.channels() == 3) {
    cv::cvtColor(img, img_gray, cv::COLOR_BGR2GRAY);
  } else {
    img_gray = img;
  }

  if (prev_image_.empty() || prev_keypoints_.size() < 100) {
    // Feature Detection
    detector_->detect(img_gray, prev_keypoints_);
    
    if (prev_keypoints_.empty()) {
      RCLCPP_WARN(get_logger(), "No features detected.");
      return;
    }

    prev_image_ = img_gray.clone();
    return;
  }

  std::vector<cv::Point2f> prev_points, curr_points;
  for (const auto & kp : prev_keypoints_) {
    prev_points.push_back(kp.pt);
  }

  std::vector<uchar> status;
  std::vector<float> error;

  // Calculate flow
  cv::calcOpticalFlowPyrLK(prev_image_, img_gray, prev_points, curr_points, status, error);

  // Filter good matches based on status and error
  std::vector<cv::Point2f> good_prev_points, good_curr_points;
  for (size_t i = 0; i < curr_points.size(); i++) {
    // Check if tracked successfully AND error is acceptable
    if (i < status.size() && status[i] && error[i] < 12.0) {
      good_prev_points.push_back(prev_points[i]);
      good_curr_points.push_back(curr_points[i]);
    }
  }

  // Visualization
  cv::Mat img_match;
  img.copyTo(img_match);
  for (size_t i = 0; i < good_curr_points.size(); i++) {
    cv::circle(img_match, good_curr_points[i], 2, cv::Scalar(0, 250, 0), 2);
    cv::line(img_match, good_prev_points[i], good_curr_points[i], cv::Scalar(0, 250, 0));
  }

  // Publish the matches image
  auto msg_matches = cv_bridge::CvImage(std_msgs::msg::Header(), "rgb8", img_match).toImageMsg();
  msg_matches->header.frame_id = image_msg->header.frame_id;
  matches_pub_.publish(*msg_matches);

  cv::Mat R, t;
  estimatePose(good_curr_points, good_prev_points, R, t);
  RCLCPP_INFO_STREAM(get_logger(), "Estimated Rotation:\n" << R << "\nEstimated Translation:\n" << t);

  prev_image_ = img_gray.clone();
  prev_keypoints_.clear();
  for (const auto & pt : good_curr_points) {
    prev_keypoints_.push_back(cv::KeyPoint(pt, 1.0f));
  }
}

void OpticalFlow::estimatePose(
  const std::vector<cv::Point2f> & curr_points,
  const std::vector<cv::Point2f> & prev_points,
  cv::Mat & R, cv::Mat & t)
{
  if (curr_points.size() < 5) {
    RCLCPP_WARN(get_logger(), "Not enough points to estimate pose.");
    return;
  }

  // Calculate the Essential Matrix
  cv::Mat essential_matrix = cv::findEssentialMat(prev_points, curr_points, K_);
  if (essential_matrix.empty() || essential_matrix.rows != 3 || essential_matrix.cols != 3) {
    RCLCPP_WARN(get_logger(), "Failed to compute Essential Matrix.");
    return;
  }

  cv::recoverPose(essential_matrix, prev_points, curr_points, K_, R, t);
}
}  // namespace bumperbot_vision

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<bumperbot_vision::OpticalFlow>("optical_flow"));
  rclcpp::shutdown();
  return 0;
}

File: bumperbot_vision/bumperbot_vision/optical_flow.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import Image, CameraInfo
from cv_bridge import CvBridge
import numpy as np

import cv2
import message_filters

class OpticalFlow(Node):
    def __init__(self):
        super().__init__('optical_flow')

        self.prev_image = None
        self.prev_keypoints = None

        self.bridge = CvBridge()

        # Feature detection and matching
        self.detector = cv2.GFTTDetector_create(
            maxCorners=500, qualityLevel=0.01, minDistance=20.0)

        # Subscribers
        image_sub = message_filters.Subscriber(
            self, Image, "left_camera/image_raw", qos_profile=qos_profile_sensor_data)
        info_sub = message_filters.Subscriber(
            self, CameraInfo, "left_camera/camera_info", qos_profile=qos_profile_sensor_data)

        self.approximate_sync = message_filters.ApproximateTimeSynchronizer(
            [image_sub, info_sub], queue_size=10, slop=0.2)
        self.approximate_sync.registerCallback(self.image_callback)

        self.matches_pub = self.create_publisher(
            Image, "matches", qos_profile_sensor_data)

    def image_callback(self, image_msg, info_msg):
        # Camera intrinsic parameters
        self.K = np.array([
            [info_msg.k[0], info_msg.k[1], info_msg.k[2]],
            [info_msg.k[3], info_msg.k[4], info_msg.k[5]],
            [info_msg.k[6], info_msg.k[7], info_msg.k[8]]
        ], dtype=np.float64)

        # Convert ROS Image to OpenCV
        try:
            current_image = self.bridge.imgmsg_to_cv2(image_msg, "bgr8")
            current_image_gray = cv2.cvtColor(current_image, cv2.COLOR_BGR2GRAY)
        except Exception as e:
            self.get_logger().error(f"Failed to convert image: {e}")
            return
        
        if self.prev_image is None or self.prev_points is None or len(self.prev_points) < 100:
            # Feature detection
            curr_keypoints = self.detector.detect(current_image_gray, None)
            
            if not curr_keypoints:
                self.get_logger().warn("No features detected to track.")
                return

            self.prev_points = np.array([kp.pt for kp in curr_keypoints], dtype=np.float32).reshape(-1, 1, 2)
            self.prev_image = current_image_gray.copy()
            return

        curr_points, status, error = cv2.calcOpticalFlowPyrLK(
            self.prev_image, current_image_gray, self.prev_points, None)
        
        # Filter good matches
        good_prev_points = []
        good_curr_points = []

        if curr_points is not None:
            for i, (p_curr, p_prev, s, e) in enumerate(zip(curr_points, self.prev_points, status, error)):
                if s[0] == 1 and e[0] < 12.0:
                    good_prev_points.append(p_prev[0])
                    good_curr_points.append(p_curr[0])

        img_match = current_image.copy()
        for pt_curr, pt_prev in zip(good_curr_points, good_prev_points):
            pt1 = (int(pt_curr[0]), int(pt_curr[1]))
            pt2 = (int(pt_prev[0]), int(pt_prev[1]))
            cv2.circle(img_match, pt1, 2, (0, 250, 0), 2)
            cv2.line(img_match, pt2, pt1, (0, 250, 0))

        # Publish Matches
        try:
            msg_matches = self.bridge.cv2_to_imgmsg(img_match, "bgr8")
            msg_matches.header.frame_id = image_msg.header.frame_id
            msg_matches.header.stamp = image_msg.header.stamp
            self.matches_pub.publish(msg_matches)
        except Exception as e:
            self.get_logger().error(f"Failed to publish matches: {e}")

        # Estimate Pose
        R, t = self.estimatePose(good_curr_points, good_prev_points)
        if R is not None and t is not None:
            self.get_logger().info(f"Estimated Rotation:\n{R}\nEstimated Translation:\n{t}")

        self.prev_image = current_image_gray.copy()
        if len(good_curr_points) > 0:
            self.prev_points = np.array(good_curr_points, dtype=np.float32).reshape(-1, 1, 2)
        else:
            self.prev_points = None

    def estimatePose(self, curr_points, prev_points):
        if len(curr_points) < 5:
            return None, None

        pts1 = np.array(prev_points, dtype=np.float32)
        pts2 = np.array(curr_points, dtype=np.float32)

        # Calculate the Essential Matrix
        essential_matrix, _ = cv2.findEssentialMat(pts1, pts2, self.K)
        if essential_matrix is None or essential_matrix.shape != (3, 3):
            return None, None
        
        # Recover Pose
        _, R, t, _ = cv2.recoverPose(essential_matrix, pts1, pts2, self.K)
        return R, t

def main(args=None):
    rclpy.init(args=args)
    optical_flow = OpticalFlow()
    rclpy.spin(optical_flow)
    optical_flow.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
